### PR TITLE
feat(Accounting): Bulk statement of accounts

### PR DESCRIPTION
**Pull request by abhishekbalam**
# Bulk Statement Of Accounts (Name not confirmed yet)

image

image

image

image

image

image

image

WIP: Please Don't Merge, some testing left.

### CODE REVIEW

**Comment by abhishekbalam** [comment]
@deepeshgarg007 

Please go through it and meanwhile, I will test and make changes.

Also for the CC field, is there is a need for a link field with User?
Can I just put, "CC me" instead? Taking the session user email id.

Asking because in Administrator role, you cannot see an entry in
the user list for it. There is something I have missed here, if we
are keeping this, do let me know what to change.

All customers with contacts which have "is_billing_contact" selected
are only considered. Primary email is optional, but if the check for it
is selected it becomes mandatory. Also if the contact does not have at
least 1 primary email in the email list of Contact it won't be selected.

When selecting the group nodes in the Fetch Customers part, all child nodes
are selected acc to lft and rgt values.

Filter duration is in Months (how old reports do you want).
There is now error checking at every stage to prevent 
monkey usage, but needs testing.

**Comment by abhishekbalam** [comment]
@deepeshgarg007 @nabinhait 

Need a few suggestions :
- Doctype Name, the present one is too long
- What should be in the Email Subject and Body
- Some fieldnames need changing like "Always send to primary contact" 
- Had to use 2 childtables for multiselect field. Is there a way to reuse them? I had tried to change property of child table but wasnt able to select one individually
- PDF format changes, style improvements
- Should we use a check to see the option to fetch customers?

Also pls point out if there are any other changes.
The code needs optimisation but it works for now.
Testing meanwhile.

File: erpnext/accounts/doctype/bsoa_cost_center_field/bsoa_cost_center_field.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+   "options": "Cost Center",
+   "reqd": 1
+  }
+ ],
+ "istable": 1,
+ "links": [],
+ "modified": "2020-07-23 21:22:51.137234",
+ "modified_by": "Administrator",
+ "module": "Accounts",
+ "name": "BSOA Cost Center Field",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bsoa_cost_center_field/bsoa_cost_center_field.json:26]: Rename to "BSOA Cost Center"

File: erpnext/accounts/doctype/bsoa_project_field/bsoa_project_field.json [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+   "options": "Project",
+   "reqd": 1
+  }
+ ],
+ "istable": 1,
+ "links": [],
+ "modified": "2020-07-23 21:19:50.560112",
+ "modified_by": "Administrator",
+ "module": "Accounts",
+ "name": "BSOA Project Field",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bsoa_project_field/bsoa_project_field.json:26]: Rename to "BSOA Project"

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bsoa.html [hunk 3]
```diff
@@ -0,0 +1,105 @@
+<h2 class="text-center" style="page-break-before:always">{{ filters.party[0] }}</h2>
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bsoa.html:1]: Rename the file to `bulk_statement_of_accounts.html`

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bsoa.html [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+<h2 class="text-center">{{ _("Statement of Account") }}</h2>
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bsoa.html:2]:
```suggestion
<h2 class="text-center">{{ _("Statement of Accounts") }}</h2>
```

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+   "label": "Filter Duration (Months)"
+  }
+ ],
+ "links": [],
+ "modified": "2020-07-25 15:09:27.285400",
+ "modified_by": "Administrator",
+ "module": "Accounts",
+ "name": "Bulk Statement Of Accounts",
+ "owner": "Administrator",
+ "permissions": [
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json:255]: Give permissions to Accounts User and Accounts Manager. And remove permissions for Administrator and System Manager.

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+   "fieldname": "filter_duration",
+   "fieldtype": "Int",
+   "label": "Filter Duration (Months)"
+  }
+ ],
+ "links": [],
+ "modified": "2020-07-25 15:09:27.285400",
+ "modified_by": "Administrator",
+ "module": "Accounts",
+ "name": "Bulk Statement Of Accounts",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json:253]: Rename to "Process Statement of Accounts", like "Process Deferred Revenue"/"Process Loan Interest Accrual"

File: erpnext/accounts/doctype/bulk_statement_of_accounts_customers/bulk_statement_of_accounts_customers.json [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  {
+   "fieldname": "customer",
+   "fieldtype": "Link",
+   "in_list_view": 1,
+   "label": "Customer",
+   "options": "Customer",
+   "reqd": 1
+  },
+  {
+   "fetch_from": "this.customers",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts_customers/bulk_statement_of_accounts_customers.json:22]: What "this.customers" does???

File: erpnext/accounts/doctype/bulk_statement_of_accounts_customers/bulk_statement_of_accounts_customers.json [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+   "fieldname": "billing_email",
+   "fieldtype": "Read Only",
+   "in_list_view": 1,
+   "label": "Billing Email"
+  },
+  {
+   "fieldname": "primary_email",
+   "fieldtype": "Read Only",
+   "in_list_view": 1,
+   "label": "Primary Email"
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts_customers/bulk_statement_of_accounts_customers.json:32]: Let's rename to Primary Contact Email

File: erpnext/accounts/doctype/bulk_statement_of_accounts_customers/bulk_statement_of_accounts_customers.json [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  }
+ ],
+ "istable": 1,
+ "links": [],
+ "modified": "2020-07-25 13:08:56.220460",
+ "modified_by": "Administrator",
+ "module": "Accounts",
+ "name": "Bulk Statement Of Accounts Customers",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts_customers/bulk_statement_of_accounts_customers.json:40]: We always use the singular form in doctype naming.

```suggestion
 "name": "Bulk Statement Of Accounts Customer",
```

File: erpnext/accounts/general_ledger.py [hunk 10]
```diff
@@ -3,345 +3,511 @@
 
 from __future__ import unicode_literals
 import frappe, erpnext
-from frappe.utils import flt, cstr, cint, comma_and, today, getdate, formatdate, now
```
Comment (nabinhait) [inline erpnext/accounts/general_ledger.py]: This is **serious** damage by mistake.

You have replaced `erpnext/accounts/general_ledger.py` by `erpnext/accounts/report/general_ledger/general_ledger.py`

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+  "cc_to",
+  "customer_collection",
+  "column_break_6",
+  "primary_mandatory",
+  "section_break_27",
+  "collection_name",
+  "column_break_29",
+  "fetch_customers",
+  "column_break_17",
+  "customer_list"
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json:43]: We generally use the child table fieldname as the plural form of most important object the child table holds like items, taxes etc. In this case, it should be "customers".

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+  "start_date",
+  "section_break_3",
+  "cc_to",
+  "customer_collection",
+  "column_break_6",
+  "primary_mandatory",
+  "section_break_27",
+  "collection_name",
+  "column_break_29",
+  "fetch_customers",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json:41]: On the right side, the button seems like hanging. Move it to the left column.

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+  "to_date",
+  "finance_book",
+  "currency",
+  "project",
+  "preferences",
+  "include_ageing",
+  "ageing_based_on",
+  "section_break_14",
+  "orientation",
+  "section_break_1",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json:26]: Move "Print Preferences" and "Auto Email" section to the end, after the customers section. The reason is GL Filters and Customers are similar fields.

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+   "fieldname": "group_by",
+   "fieldtype": "Select",
+   "label": "Group By",
+   "options": "\nGroup by Voucher\nGroup by Voucher (Consolidated)"
+  },
+  {
+   "fieldname": "currency",
+   "fieldtype": "Select",
+   "label": "Currency",
+   "options": "\nAED\nAUD\nCHF\nCNY\nEUR\nGBP\nINR\nJPY\nUSD"
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json:173]: Why hard-coded? Make it a Link field to Currency.

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.js [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+						else{
+							window.location = url;
+						}
+					},
+					fail: function(){
+						frappe.msgprint('No Records for this day');
+					}
+				});
+			});
+			// frm.add_custom_button('Trigger Auto Email',function(){
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.js:43]: Remove commented lines

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.js [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+		if(frm.doc.frequency != ''){
+			frm.set_value('start_date', frappe.datetime.get_today());
+		}
+		else{
+			frm.set_value('start_date', '');
+		}
+	},
+	fetch_customers: function(frm){
+		if(frm.doc.collection_name){
+			console.log(frm.doc.primary_mandatory)
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.js:78]: remove console.log

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.js [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+							for (const customer of r.message){
+								var row = frm.add_child('customer_list');
+								row.customer = customer.name;
+								row.primary_email = customer.primary_email;
+								row.billing_email = customer.billing_email;
+							}
+							frm.refresh_field('customer_list');
+						}
+						else{
+							frappe.msgprint('No Customers found with selected options!');
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.js:100]: We generally don't use exclamation in messages

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.js [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+		frappe.call({
+			method: 'erpnext.accounts.doctype.bulk_statement_of_accounts.bulk_statement_of_accounts.get_customer_emails',
+			args: {
+				'customer_name': row.customer,
+				'primary_mandatory': frm.doc.primary_mandatory
+			},
+			callback: function(r){
+				if(!r.exe){
+					if(r.message.length){
+						console.log(r.message)
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.js:127]: remove

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.py [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+		result = ''.join(list(statement_dict.values()))
+		return get_pdf(result, {'orientation': doc.orientation})
+	else:
+		for customer, statement_html in statement_dict.items():
+			statement_dict[customer]=get_pdf(statement_html, {'orientation': doc.orientation})
+		return statement_dict
+
+@frappe.whitelist()
+def fetch_customers(customer_collection, collection_name, primary_mandatory):
+	print(type(primary_mandatory))
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.py:91]: remove print

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.py [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+	elif consolidated:
+		result = ''.join(list(statement_dict.values()))
+		return get_pdf(result, {'orientation': doc.orientation})
+	else:
+		for customer, statement_html in statement_dict.items():
+			statement_dict[customer]=get_pdf(statement_html, {'orientation': doc.orientation})
+		return statement_dict
+
+@frappe.whitelist()
+def fetch_customers(customer_collection, collection_name, primary_mandatory):
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.py:90]: split into smaller functions. Each customer_collection can be a separate function.

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.py [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+def get_customer_emails(customer_name, primary_mandatory, billing_and_primary=True):
+	billing_email = frappe.db.sql("""
+		SELECT c.email_id FROM `tabContact` AS c JOIN `tabDynamic Link` AS l ON c.name=l.parent \
+		WHERE l.link_doctype='Customer' and l.link_name='""" + customer_name + """' and \
+		c.is_billing_contact=1 \
+		order by c.creation desc""")
+
+	if len(billing_email) == 0 or (billing_email[0][0] is None):
+		if billing_and_primary:
+			frappe.throw('No billing email found for customer: '+ customer_name)
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.py:148]: If no billing email, but there is a primary email, don't throw error, send the email to primary contact.

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.py [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+				if clist.customer == customer:
+					recipients.append(clist.billing_email)
+					if doc.primary_mandatory:
+						recipients.append(clist.primary_email or '')
+			cc = None
+			if doc.cc_to != '':
+				try:
+					cc=frappe.get_doc('User', doc.cc_to, 'email')
+				except:
+					pass
```
Comment (nabinhait) [inline erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.py]: Make this a separate function `get_recipients` and return recipients and cc.

**Review by nabinhait — CHANGES_REQUESTED** [review]
- Are there any changes in purchase_invoice.json, purchase_order.json, salary_slip.json, purchase_receipt.json, serial_no.js and serial_no.json? If not, please checkout them.

**Comment by nabinhait** [comment]
> * Doctype Name, the present one is too long

**Process Statement of Accounts**

> * What should be in the Email Subject and Body

**Add 2 fields to take input of subject and body message like "Notification" doctype. The fields should support jinja as well.**

> * Some fieldnames need changing like "Always send to primary contact"

**"Send to primary contact"**

> * Had to use 2 childtables for multiselect field. Is there a way to reuse them? I had tried to change property of child table but wasnt able to select one individually

**Keep as it is** 

> * PDF format changes, style improvements

**Consult the framework team**

> * Should we use a check to see the option to fetch customers?

**No, the button is ok, but move to left side**

File: erpnext/accounts/doctype/bulk_statement_of_accounts_customers/bulk_statement_of_accounts_customers.json (hunk 7 above)
Comment (abhishekbalam) [reply erpnext/accounts/doctype/bulk_statement_of_accounts_customers/bulk_statement_of_accounts_customers.json:22]: It was to test something.
Removed now

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.js (hunk 17 above)
Comment (abhishekbalam) [reply erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.js:100]: noted

File: erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json (hunk 14 above)
Comment (abhishekbalam) [reply erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.json:173]: Done, also added a filter for enabled

**Comment by abhishekbalam** [comment]
@nabinhait @deepeshgarg007 
My git history is too broken to remove conflicts, started a new branch with the requested changes.

New PR Link:
https://github.com/frappe/erpnext/pull/22901

Sorry for the inconvenience.
Will close this soon.

**Pull request closed without merging**